File: Coding_Questions/C++/BinarySearch.cpp
// ---Binary Search---

#include <vector>
#include "math.h"


int binary_search_helper(std::vector<int> array, int target, int lower_pointer, int right_pointer) {
    while (lower_pointer <= right_pointer) {
        int middle_point = floor((lower_pointer + right_pointer) / 2);
        if (array[middle_point] == target) {
            return middle_point;
        } else if (array[middle_point] > target) {
            right_pointer = middle_point - 1;
        } else if (array[middle_point] < target) {
            lower_pointer = middle_point + 1;
        }
    }
    return -1;
}


int binary_search(std::vector<int> array, int target) {
    return binary_search_helper(array, target, 0, array.size());
}

File: Coding_Questions/JavaScript/BinarySearch.js
// ---Binary Search---

const binarySearch = (array, target) => {
  return binarySearchHelper(array, target, 0, array.length - 1);
};

const binarySearchHelper = (array, target, leftPointer, rightPointer) => {
  while (leftPointer <= rightPointer) {
    const middlePoint = Math.floor((leftPointer + rightPointer) / 2);
    if (array[middlePoint] === target) {
      return middlePoint;
    } else if (array[middlePoint] > target) {
      rightPointer = middlePoint - 1;
    } else if (array[middlePoint] < target) {
      leftPointer = middlePoint + 1;
    }
  }
  return -1;
};

console.log(binarySearch([0, 1, 21, 33, 45, 45, 61, 71, 72, 73], 33)); // 3
